lib/cocoapods/generator/xcconfig/aggregate_xcconfig.rb

```diff
@@ -103,9 +102,15 @@ def generate_settings_to_import_pod_targets
             end
             build_settings = {
               'PODS_FRAMEWORK_BUILD_PATH' => target.scoped_configuration_build_dir,
-              # Make headers discoverable by `import "…"`
-              'OTHER_CFLAGS' => '$(inherited) ' + XCConfigHelper.quote(header_search_paths, '-iquote'),
+              # Make framework headers discoverable by `import "…"`
+              'OTHER_CFLAGS' => '$(inherited) ' + XCConfigHelper.quote(framework_header_search_paths, '-iquote'),
             }
+            if pod_targets.any? { |t| !t.should_build? }
+              # Make library headers discoverale by `#import "…"`
+              library_header_search_paths = target.sandbox.public_headers.search_paths(target.platform)
```

Comment:
FYI, #3868 attempts to fix the excess of public headers. I think it fits nicely into its own PR.